compiler/rustc_passes/src/custom_abi.rs

```diff
@@ -0,0 +1,153 @@
+                tcx.hir_body(*body_id)
+            }
+            hir::Node::TraitItem(hir::TraitItem {
+                kind: hir::TraitItemKind::Fn(sig, trait_fn),
+                ..
+            }) => match trait_fn {
+                hir::TraitFn::Required(_) => {
+                    check_signature(tcx, def_id, sig, false);
+                    continue;
+                }
+                hir::TraitFn::Provided(body_id) => {
+                    check_signature(tcx, def_id, sig, true);
+                    tcx.hir_body(*body_id)
+                }
+            },
+            _ => continue,
+        };
+
+        let mut visitor = CheckCustomAbi { tcx };
+        visitor.visit_body(body);
+    }
+}
+
+fn check_signature<'tcx>(tcx: TyCtxt<'tcx>, def_id: LocalDefId, sig: &FnSig<'tcx>, has_body: bool) {
+    if sig.header.abi == ExternAbi::Custom {
+        // Function definitions that use `extern "custom"` must be naked functions.
+        if has_body && !tcx.has_attr(def_id, sym::naked) {
+            tcx.dcx().emit_err(AbiCustomClothedFunction {
+                span: sig.span,
+                naked_span: sig.span.shrink_to_lo(),
+            });
+        }
+
+        // Function definitions that use `extern "custom"` must unsafe.
+        if sig.header.is_safe() {
+            tcx.dcx().emit_err(AbiCustomSafeFunction {
+                span: sig.span,
+                unsafe_span: sig.span.shrink_to_lo(),
+            });
+        }
+    }
+}
+
+struct CheckCustomAbi<'tcx> {
+    tcx: TyCtxt<'tcx>,
+}
+
+impl<'tcx> Visitor<'tcx> for CheckCustomAbi<'tcx> {
+    type NestedFilter = OnlyBodies;
+
+    fn maybe_tcx(&mut self) -> Self::MaybeTyCtxt {
+        self.tcx
+    }
+
+    fn visit_expr(&mut self, expr: &'tcx hir::Expr<'tcx>) {
+        let opt_span_and_abi = match expr.kind {
+            ExprKind::Call(fun, _) => {
+                let fun_ty = self.tcx.typeck(fun.hir_id.owner.def_id).node_type(fun.hir_id);
+
+                match *fun_ty.kind() {
+                    ty::FnDef(def_id, _) => {
+                        let sig = self.tcx.fn_sig(def_id).skip_binder().skip_binder();
+                        Some((expr.span, sig.abi))
+                    }
+                    ty::FnPtr(_, header) => Some((expr.span, header.abi)),
+                    _ => None,
+                }
+            }
+
+            ExprKind::MethodCall(_, receiver, _, span) => {
+                let opt_def_id = self
+                    .tcx
+                    .typeck(receiver.hir_id.owner.def_id)
+                    .type_dependent_def_id(expr.hir_id);
+
+                opt_def_id.map(|def_id| {
+                    let sig = self.tcx.fn_sig(def_id).skip_binder().skip_binder();
+                    (span, sig.abi)
+                })
+            }
+            _ => None,
+        };
+
+        if let Some((span, ExternAbi::Custom)) = opt_span_and_abi {
+            self.tcx.dcx().emit_err(AbiCustomCall { span });
+        }
```

Comment:
From https://github.com/rust-lang/rust/issues/140566#issuecomment-2846205457 and the few messages following, it sounds like there are a couple of other ABIs that should be uncallable (specifically interrupt ABIs). Maybe this should go via an `is_abi_callable` function next to `is_abi_supported` in `rustc_target`?